### PR TITLE
feat(x/swaprouter, x/gamm): move swap messages from gamm to swaprouter

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #3064

## What is the purpose of the change

This PR completely moves `SwapExactAmountIn` and  `SwapExactAmountOut` from `gamm` into the new `swaprouter` SDK module so that we have a unified generalized API for swaps.

All pre-existing CLI, query logic and tests has been ported. As of right now, the new messages simply route the calls to gamm.
No new tests have been added.

Relevant `wasmbindings` are completely removes since the goal is to continue using stargate query instead of bindings,

Connecting `swaprouter` to the `concentrated-liqudity` module should happen next in:
https://github.com/osmosis-labs/osmosis/issues/3063

## Simulation Hack

Pre-existing setup for simulation requires access to some gamm keeper APIs.

Current simulation utility functions do not provide support to wire additional keepers. As a result, extra keeper methods were defined on `swaprouter` keeper to route the calls to `gamm` keeper internally.
https://github.com/osmosis-labs/osmosis/blob/ba30351c01c250d2aa3f850fcd9bec621c8b1c8b/x/swaprouter/types/routes.go#L29-L38

We should refactor simulator API to provide wiring additional keepers to avoid this issue.

This is to be fixed in #3087

## Testing and Verifying

This change is a trivial rework / code cleanup without any test coverage.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/concentrated-liquidity/swaps.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// TODO: godoc
+func (k Keeper) MultihopSwapExactAmountOut(
+	ctx sdk.Context,
+	sender sdk.AccAddress,
+	routes []swaproutertypes.SwapAmountOutRoute,
+	tokenInMaxAmount sdk.Int,
+	tokenOut sdk.Coin,
+) (tokenInAmount sdk.Int, err error) {
+	return sdk.Int{}, nil
+}
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/swaps.go:29]: Do we support multihop on cl pools for mvp?

File: x/gamm/client/cli/tx.go [hunk 2]
```diff
@@ -146,66 +143,6 @@ func NewExitPoolCmd() *cobra.Command {
 	return cmd
 }
 
-func NewSwapExactAmountInCmd() *cobra.Command {
```
Comment (czarcas7ic) [inline x/gamm/client/cli/tx.go:149]: Do you for-see the cli utilizing swap router module directly for the swap? In my mind I thought that we would still support swap calls from gamm that would get sent to the router with the required fields pre-filled. We dont have to address this now, just wanted to bring this up.

File: x/gamm/keeper/grpc_query.go [hunk 3]
```diff
@@ -18,20 +17,6 @@ import (
...
-	maxInt = maxInt.Exp(maxInt, big.NewInt(256), nil)
-
-	_sdkIntMaxValue, ok := sdk.NewIntFromString(maxInt.Sub(maxInt, big.NewInt(1)).String())
-	if !ok {
-		panic("Failed to calculate the max value of sdk.Int")
-	}
-
-	sdkIntMaxValue = _sdkIntMaxValue
-}
-
```
Comment (czarcas7ic) [inline x/gamm/keeper/grpc_query.go:34]: this seems unrelated?

File: x/gamm/types/errors.go [hunk 4]
```diff
@@ -29,19 +29,18 @@ var (
...
+	ErrNegativeExitFee          = sdkerrors.Register(ModuleName, 23, "exit fee is negative")
+	ErrTooMuchSwapFee           = sdkerrors.Register(ModuleName, 24, "swap fee should be lesser than 1 (100%)")
+	ErrTooMuchExitFee           = sdkerrors.Register(ModuleName, 25, "exit fee should be lesser than 1 (100%)")
+	ErrNotPositiveWeight        = sdkerrors.Register(ModuleName, 26, "token weight should be greater than 0")
+	ErrWeightTooLarge           = sdkerrors.Register(ModuleName, 27, "user specified token weight should be less than 2^20")
+	ErrNotPositiveCriteria      = sdkerrors.Register(ModuleName, 28, "min out amount or max in amount should be positive")
+	ErrNotPositiveRequireAmount = sdkerrors.Register(ModuleName, 29, "required amount should be positive")
+	ErrTooManyTokensOut         = sdkerrors.Register(ModuleName, 30, "tx is trying to get more tokens out of the pool than exist")
+	ErrSpotPriceOverflow        = sdkerrors.Register(ModuleName, 31, "invalid spot price (overflowed)")
+	ErrSpotPriceInternal        = sdkerrors.Register(ModuleName, 32, "internal spot price error")
```
Comment (czarcas7ic) [inline x/gamm/types/errors.go:43]: It seems like we are assigning errors to previously assigned error numbers, which AFAIK we should not do

**Review by czarcas7ic — APPROVED** [review]

File: x/concentrated-liquidity/swaps.go (hunk 1 above)
Comment (p0mvn) [reply x/concentrated-liquidity/swaps.go:29]: Yes, should be done in a separate issue https://github.com/osmosis-labs/osmosis/issues/3063

File: x/gamm/client/cli/tx.go (hunk 2 above)
Comment (p0mvn) [reply x/gamm/client/cli/tx.go:149]: I think all swap entrypoints should be inside this module. Not sure why we would want to fragment CLI / messages / queries across various modules.

File: x/gamm/keeper/grpc_query.go (hunk 3 above)
Comment (p0mvn) [reply x/gamm/keeper/grpc_query.go:34]: Confirmed that It is used by swap query tests and nothing else

File: x/gamm/types/errors.go (hunk 4 above)
Comment (p0mvn) [reply x/gamm/types/errors.go:43]: I think this concern is valid across patches due to state-compatibility concerns.

Since this is a major upgrade, should be fine. 

I will ask one more reviewer to confirm if that's the case

Comment (ValarDragon) [reply x/gamm/types/errors.go:43]: I prefer we don't do any more sdkerrors.register anywhere in the codebase.

Comment (p0mvn) [reply x/gamm/types/errors.go:43]: Yeah I only removed an sdk error here and ported it as Go error to the new `swaprouter` module. I did not add any new sdkerror.

However, removing an sdkerror with an ID in the middle, made me have to lower all remaining IDs by 1.

I'm wondering if that is allowed?

**Comment by p0mvn** [comment]
R4R. e2e seems to be failing on `concentrated-liquidity-main`. Created https://github.com/osmosis-labs/osmosis/issues/3089 to investigate

File: x/swaprouter/client/query_proto_wrap.go [hunk 5]
```diff
@@ -13,9 +18,91 @@ type Querier struct {
...
+	maxInt = maxInt.Exp(maxInt, big.NewInt(256), nil)
+
+	_sdkIntMaxValue, ok := sdk.NewIntFromString(maxInt.Sub(maxInt, big.NewInt(1)).String())
+	if !ok {
+		panic("Failed to calculate the max value of sdk.Int")
+	}
+
+	sdkIntMaxValue = _sdkIntMaxValue
+}
+
```
Comment (mattverse) [inline x/swaprouter/client/query_proto_wrap.go:34]: this is used in tests right? should we move it to tests?

File: x/swaprouter/module/module.go [hunk 6]
```diff
@@ -124,9 +133,18 @@ func (AppModule) BeginBlock(_ sdk.Context, _ abci.RequestBeginBlock) {}
...
 func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.ValidatorUpdate {
-	// am.k.EndBlock(ctx)
 	return []abci.ValidatorUpdate{}
 }
 
 // ConsensusVersion implements AppModule/ConsensusVersion.
 func (AppModule) ConsensusVersion() uint64 { return 1 }
+
+// **** simulation implementation ****
+// GenerateGenesisState creates a randomized GenState of the gamm module.
```
Comment (mattverse) [inline x/swaprouter/module/module.go:143]:
```suggestion
// GenerateGenesisState creates a randomized GenState of the swaprouter module.
```

File: x/swaprouter/types/routes.go [hunk 7]
```diff
@@ -1,5 +1,80 @@
...
+		routes []SwapAmountOutRoute,
+		tokenInMaxAmount sdk.Int,
+		tokenOut sdk.Coin,
+	) (tokenInAmount sdk.Int, err error)
+}
+
+// SimulationExtension defines the swap simulation extension.
+// TODO: refactor simulator setup logic to avoid having to define these
+// extra methods just for the simulation.
+type SimulationExtension interface {
```
Comment (mattverse) [inline x/swaprouter/types/routes.go:32]: Why not name it SwapExtension?
I personally think the naming simulation makes it consider as a simulation test extension

File: x/swaprouter/keeper.go [hunk 8]
```diff
@@ -12,21 +13,21 @@ type Keeper struct {
 	storeKey sdk.StoreKey
 	// TODO: remove nolint
 	// nolint: unused
-	gammKeeper types.SwapI
+	gammKeeper types.SimulationExtension
```
Comment (mattverse) [inline x/swaprouter/keeper.go:16]: same question, how about SwapExtension or GammSwapExtension?

File: app/upgrades/v13/upgrades.go [hunk 9]
```diff
@@ -18,6 +19,7 @@ func CreateUpgradeHandler(
 ) upgradetypes.UpgradeHandler {
 	return func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
 		keepers.LockupKeeper.SetParams(ctx, lockuptypes.DefaultParams())
+		keepers.SwapRouterKeeper.SetParams(ctx, swaproutertypes.DefaultParams())
```
[the remark this answers is not in the document]
Comment (mattverse) [reply app/upgrades/v13/upgrades.go:22]: Nevermind, we need this for the e2e test right?

File: x/swaprouter/client/cli/cli_test.go [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+	testCases := []struct {
+		name string
+		args []string
+
+		expectErr    bool
+		respType     proto.Message
+		expectedCode uint32
+	}{
+		{
+			"swap exact amount out", // osmosisd tx gamm swap-exact-amount-out 10stake 20 --swap-route-pool-ids=1 --swap-route-denoms=node0token --from=validator --keyring-backend=test --chain-id=testing --yes
```
Comment (czarcas7ic) [inline x/swaprouter/client/cli/cli_test.go:91]: nit: for the comment, this is now via swaprouter and not gamm right?

File: x/swaprouter/client/cli/cli_test.go [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+	testCases := []struct {
+		name string
+		args []string
+
+		expectErr    bool
+		respType     proto.Message
+		expectedCode uint32
+	}{
+		{
+			"swap exact amount in", // osmosisd tx gamm swap-exact-amount-in 10stake 3 --swap-route-pool-ids=1 --swap-route-denoms=node0token --from=validator --keyring-backend=test --chain-id=testing --yes
```
Comment (czarcas7ic) [inline x/swaprouter/client/cli/cli_test.go:237]: same

**Review by czarcas7ic — APPROVED** [review]

File: app/upgrades/v13/upgrades.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply app/upgrades/v13/upgrades.go:22]: yeah I would keep this for now, and remove if we decide that params are unneded

File: wasmbinding/message_plugin.go [hunk 12]
```diff
@@ -203,75 +199,6 @@ func PerformBurn(f *tokenfactorykeeper.Keeper, ctx sdk.Context, contractAddr sdk
 	return nil
 }
 
-// swapTokens swaps one denom for another.
-func (m *CustomMessenger) swapTokens(ctx sdk.Context, contractAddr sdk.AccAddress, swap *bindings.SwapMsg) ([]sdk.Event, [][]byte, error) {
-	_, err := PerformSwap(m.gammKeeper, ctx, contractAddr, swap)
-	if err != nil {
-		return nil, nil, sdkerrors.Wrap(err, "perform swap")
-	}
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply wasmbinding/message_plugin.go:211]: Bindings are deprecated

File: x/swaprouter/keeper.go (hunk 8 above)
Comment (p0mvn) [reply x/swaprouter/keeper.go:16]: This is needed for simulator only. I created an issue to refactor this later

File: x/swaprouter/client/query_proto_wrap.go (hunk 5 above)
Comment (p0mvn) [reply x/swaprouter/client/query_proto_wrap.go:34]: This is also used in `EstimateSwapExactAmountOut`

File: x/swaprouter/types/routes.go (hunk 7 above)
Comment (p0mvn) [reply x/swaprouter/types/routes.go:32]: This is a simulation test extension
